src/YieldDistributor.sol

```diff
@@ -242,21 +240,36 @@ contract YieldDistributor is OwnableUpgradeable {
     function _castVote(address _account, uint256[] calldata _points, uint256 _votingPower) internal {
         if (_points.length != projects.length) revert IncorrectNumberOfProjects();
 
+        // This loop cannot be combined with the loop below because the total points must be calculated first
+        // to perform the correct calculation for each project
         uint256 _totalPoints;
-
         for (uint256 i; i < _points.length; ++i) {
             if (_points[i] > maxPoints) revert ExceedsMaxPoints();
             _totalPoints += _points[i];
         }
 
-        if (_totalPoints == 0) revert ZeroVotePoints();
+        bool _hasVotedInCycle = accountLastVoted[_account] > lastClaimedBlockNumber;
+        uint256[] storage _voterDistributions = voterDistributions[_account];
+        // If the user has not voted in the current cycle, initialize the voter distribution array with the correct length
+        if (!_hasVotedInCycle) {
 
+            delete voterDistributions[_account];
+            for (uint256 i; i < _points.length; ++i) {
+                _voterDistributions.push(0);
+            }
+        }
         for (uint256 i; i < _points.length; ++i) {
-            projectDistributions[i] += ((_points[i] * _votingPower * PRECISION) / _totalPoints) / PRECISION;
+            uint256 _currentProjectDistribution = ((_points[i] * _votingPower * PRECISION) / _totalPoints) / PRECISION;
+            projectDistributions[i] += _currentProjectDistribution;
+            if (_hasVotedInCycle) projectDistributions[i] -= _voterDistributions[i];
+
+            _voterDistributions[i] = _currentProjectDistribution;
         }
 
+        if (_totalPoints == 0) revert ZeroVotePoints();
```

Comment:
why move this revert down here?

Reply:
relic of mistaken refactoring I did, amended with 720b486